app/src/main/java/com/example/samplegallery/PhotoBlowupFragment.java

```diff
@@ -86,16 +99,32 @@ public void onResponse(JSONObject response) {
                         try {
                             sizes = response.getJSONObject("sizes").getJSONArray("size");
                             photoUrl = sizes
-                                    .getJSONObject(sizes.length() - 1)
+                                    // originally, you were getting the original (very large) image size, unnecessarily since it didn't even fit the screen
+                                    // for faster loading speeds, get a smaller image size
+                                    .getJSONObject(sizes.length() - 3)
                                     .getString("source");
                         } catch (JSONException e) {
                             e.printStackTrace();
                         }
 
-                        // finally set the url for the image
-                        img.setImageUrl(
-                                photoUrl,
-                                VolleyRequestQueue.getInstance(getContext()).getImgLoader());
+                        // create a request queue for the images themselves
+                        ImageRequest bitmapRequest = new ImageRequest(photoUrl,
+                                new Response.Listener<Bitmap>() {
+                                    @Override
+                                    public void onResponse(Bitmap bitmap) {
+                                        img.setImageBitmap(bitmap);
+                                        photoDescriptionView.setVisibility(View.VISIBLE);
+                                        photoTitleView.setVisibility(View.VISIBLE);
+                                        progressBar.setVisibility(View.INVISIBLE);
+                                    }
+                                }, 0, 0, null,
+                                new Response.ErrorListener() {
+                                    @Override
+                                    public void onErrorResponse(VolleyError error) {
+                                        Log.d("IMAGEREQUEST ERROR: ", error.toString());
+                                    }
+                                });
+                        VolleyRequestQueue.getInstance(getContext()).addToRequestQueue(bitmapRequest);
                     }
                 },
                 new VolleyErrorListener(getContext())
```

Comment:
While this solution to the problem of synchronization works, it is not reusable. What happens if I want to have the loading spinner on every image I will be pulling from the internet? You got rid of the networkimageview which asked you to extend -- why so?

Try fixing this by extending the NetworkImageView, which would make your life that much easier! Try doing it yourself and thinking about making this as abstract as possible (for instance, why restrict yourself to just changing the visibility? Why not allow any arbitrary callback to be executed when the image has been loaded?). If you get stuck, take a look at https://github.com/YaleSTC/android_testing_training/commit/60cf8d6f1b9add182aea541fccf910fc898146dd, but use that as the last resort :)